File: trabalho-final-poo/Livro.cpp
#include <locale>
#include "Livro.h"

Livro::Livro()
{
	this->ISBN= "NULL";
}

Livro::~Livro()
{
}

bool Livro::validarISBN(string &isbn) {
    if (isbn.length() != 13) {
        return false; // ISBN-13 deve ter exatamente 13 dígitos
    }

    int soma = 0;

    for (int i = 0; i < 12; ++i) {
        int digito = isbn[i] - '0';

        // Alterna entre multiplicar por 1 e por 3
        soma += (i % 2 == 0) ? digito : 3 * digito;
    }

    int digitoVerificador = (10 - (soma % 10)) % 10;
    return (digitoVerificador == isbn[12] - '0');
}

int Livro::getEdicao (){
	return this->edicao;
}
string Livro::getEditora (){
	return this->editora;
}
string Livro::getAutor (){
	return this->autor;
}
int Livro::getNumPags(){
	return this->numPags;
}
string Livro::getISBN (){
	return this->ISBN;
}
void Livro::setEdicao(int edicao){
	this->edicao=edicao;
}
void Livro::setEditora(string editora){
	this->editora=editora;
}
void Livro::setAutor (string autor){
	this->autor=autor;
}
void Livro::setNumPags (int numPags){
	this->numPags=numPags;
}
void Livro::setISBN (string ISBN){
	if (validarISBN(ISBN)){
		this->ISBN=ISBN;
	} else {
		cout <<"O valor da ISBN dado nao e valido, por favor inserir um valor diferente"<<endl;	
	}
}

void Livro::imprimeFichaTecnica(){
	locale::global(std::locale(""));
	int larguraCaixa = 70;
	cout  << endl << setfill('=')<< setw(larguraCaixa)<< "" << setfill(' ')<<endl;
	cout << "| Livro " << setw(larguraCaixa - 8) << " |" << endl;
	cout << "| Autor: " << setw(larguraCaixa - 11) << this->autor << " |" << endl;
	Item::imprimeFichaTecnica();
	cout <<"| Editora: " << setw(larguraCaixa - 13) << this->editora << " |" << endl;
	cout << "| "<< this->edicao <<"o Edicao"<<setw(larguraCaixa - 11) << " |" << endl;
	cout << "| "<< this->numPags <<" Paginas"<<setw(larguraCaixa - 13) << " |" << endl;
	cout <<"| ISBN: " << setw(larguraCaixa - 10) << this->ISBN << " |" << endl;
	cout << setfill('=')<< setw(larguraCaixa)<< "" << setfill(' ')<<endl;	
}

Livro* Livro::cadastra(){
	Livro *l = new Livro();
	string nome;
	
	int edicao;
	string editora;
	string autor;
	int numPags;
	string ISBN;
	int qtd;
	string tags;
	cin.ignore();
	cout<<"Insira um nome:"<<endl;
	getline (cin, nome);
	l->setNome(nome);
	cout<<"Insira a editora:"<<endl;
	getline (cin,editora);
	l->setEditora(editora);
	cout<<"Insira o numero de paginas:"<<endl;
	cin >> numPags;
	l->setNumPags(numPags);
	cout<<"Insira a edicao:"<<endl;
	cin >> edicao;
	l->setEdicao(edicao);
	cout<<"Insira o autor:"<<endl;
	getline (cin,autor);
	l->setAutor(autor);
	cout<<"Insira a quantidade:"<<endl;
	cin >>qtd;
	l->setQtd(qtd);
	
	while (l->ISBN == "NULL"){
		cout <<"Insira o codigo ISBN"<<endl;
		getline (cin, ISBN);
		l->setISBN(ISBN);
	}
	
	while (tags != "sair"){
		cin.ignore();
		cout<<"Insira uma tag: (ou escreva sair para encerrar)"<<endl;
		getline (cin, tags);
		if (tags !="sair"){
			l->addTag(tags);
		}
	}
	return l;
}
#include "Livro.h"

File: acervoBiblioteca/src/main/java/com/poo2/acervobiblioteca/itens/Livro.java
package com.poo2.acervobiblioteca.itens;

import java.util.Scanner;

public class Livro extends Item{
    private int edicao;
    private String editora;
    private String autor;
    private int numPags;
    private String ISBN = "NULL";

    public int getEdicao() {
        return edicao;
    }

    public void setEdicao(int edicao) {
        this.edicao = edicao;
    }

    public String getEditora() {
        return editora;
    }

    public void setEditora(String editora) {
        this.editora = editora;
    }

    public String getAutor() {
        return autor;
    }

    public void setAutor(String autor) {
        this.autor = autor;
    }

    public int getNumPags() {
        return numPags;
    }

    public void setNumPags(int numPags) {
        this.numPags = numPags;
    }

    public String getISBN() {
        return ISBN;
    }

    public void setISBN(String ISBN) {
        if (validarISBN(ISBN)){
            this.ISBN= ISBN;
        }else{
            System.out.println("O valor da ISBN dado nao e valido, por favor inserir um valor diferente");
        }
    }

    private boolean validarISBN(String isbn) {
        if (isbn.length() != 13) {
            return false; // ISBN-13 deve ter exatamente 13 dígitos
        }
        int soma = 0;
        for (int i = 0; i < 12; ++i) {
            int digito = isbn.charAt(i) - '0';

            // Alterna entre multiplicar por 1 e por 3
            soma += (i % 2 == 0) ? digito : 3 * digito;
        }

        int digitoVerificador = (10 - (soma % 10)) % 10;
        return (digitoVerificador == isbn.charAt(12) - '0');
    }

    @Override
    public void imprimeFichaTecnica() {
        System.out.printf("Livro:\nAutor:%s\n", this.autor);
        super.imprimeFichaTecnica();
        System.out.printf("Editora: %s, %dº edição, %d páginas\n", this.editora, this.edicao, this.numPags);
        System.out.printf("ISBN: %s\n", this.ISBN);
        System.out.println("============================================================================");

    }

    @Override
    public void cadastra() {
        super.cadastra();
        Scanner scanner = new Scanner(System.in);
        System.out.println("Insira o autor:");
        this.autor= scanner.nextLine();
        System.out.println("Insira o número de páginas: ");
        this.numPags=scanner.nextInt();
        scanner= new Scanner(System.in);
        System.out.println("Insira a editora:");
        this.editora= scanner.nextLine();
        System.out.println("Insira a edição:");
        this.edicao= scanner.nextInt();
        while (this.ISBN == "NULL"){
            System.out.println("Insira o código ISBN:");
            setISBN(scanner.next());
        }
        cadastraTags();
    }
}
